napari/types.py

```diff
@@ -124,23 +124,37 @@ def reader_function(*args, **kwargs) -> List[LayerData]:
 
 def _register_types_with_magicgui():
     """Register napari.types objects with magicgui."""
+    import sys
+    from concurrent.futures import Future
+
     from . import layers
     from .utils import _magicgui as _mgui
 
-    _mgui.register_type(
-        LayerDataTuple,
-        return_callback=_mgui.add_layer_data_tuples_to_viewer,
-    )
-    _mgui.register_type(
-        List[LayerDataTuple],
-        return_callback=_mgui.add_layer_data_tuples_to_viewer,
-    )
+    for _type in (LayerDataTuple, List[LayerDataTuple]):
+        _mgui.register_type(
+            _type,
+            return_callback=_mgui.add_layer_data_tuples_to_viewer,
+        )
+        if sys.version_info >= (3, 9):
+            _mgui.register_type(
+                Future[_type], return_callback=_mgui.add_future_data  # type: ignore
+            )
+
     for layer_name in layers.NAMES:
+        data_type = globals().get(f'{layer_name.title()}Data')
         _mgui.register_type(
-            globals().get(f'{layer_name.title()}Data'),
+            data_type,
             choices=_mgui.get_layers_data,
             return_callback=_mgui.add_layer_data_to_viewer,
         )
+        if sys.version_info >= (3, 9):
+            _mgui.register_type(
+                Future[data_type],  # type: ignore
+                choices=_mgui.get_layers_data,
+                return_callback=partial(
+                    _mgui.add_future_data, _from_tuple=False
+                ),
+            )
```

Comment:
Do we need a warning here in the `else` - like "This feature only supported for >= 3.9" or something similar, or is it not possible for 3.7/ 3.8 users to even get to this point in the code. I just don't want us to silently skip them

Reply:
it's not possible for a 3.7/3.8 user to annotate something in the first place with `Future[something]`... so not registering that subscripted future for <3.9 with magicgui is fine (they'll get the error before they get to this point)

Reply:
ok, i thought that might be the case based on your comment in main thread, but just wanted to check - thanks

Reply:
Is that error easy to understand in the context of the project or more of a general error that would be difficult to understand in practice?

Reply:
it's a more general error.  Basically, it would need to be made clear in examples and documentation (specifically, [this page](https://napari.org/guides/stable/magicgui.html)) that if you want to use `Future[something]` with magicgui and napari, you need to be on python 3.9

Reply:
Fair enough!